### PR TITLE
Recover from missed RPC events after WebSocket subscription is closed by Tendermint

### PR DESCRIPTION
**Pull request by romac**
Closes: #1196

Depends on:
- [x] https://github.com/informalsystems/tendermint-rs/pull/929
- [x] https://github.com/informalsystems/tendermint-rs/pull/931
- [x] https://github.com/informalsystems/tendermint-rs/pull/935

## Description

After some investigation, the culprit for #1196 seems to be that Tendermint is closing the WebSocket connection over which we listen for IBC events whenever more than 100 txs are included in a single block [0], as we are not able to pull the events fast enough over the WebSocket connection to avoid completely filling the event buffer in Tendermint (which currently has a hard-coded capacity of 100 events, hence the issue).

We never noticed this previously since this problem only appears in practice with a high-enough commit/propose timeout (to allow enough txs to be included in a single block), and we were testing with a lower value for the timeouts.

Now that we landed some changes in tendermint-rs [1] which allow us to notice the connection being closed, this PR makes use of this to resubscribe to the events and trigger a packet clear whenever we notice the connection being closed under our feet.

[0] https://github.com/tendermint/tendermint/issues/6729
[1] https://github.com/informalsystems/tendermint-rs/pull/929
______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: relayer/src/worker/packet.rs [hunk 1]
```diff
@@ -110,10 +110,13 @@ impl PacketWorker {
                     // once at start, and _forced_ at predefined block intervals.
                     let force_packet_clearing = self.clear_packets_interval != 0
                         && height.revision_height % self.clear_packets_interval == 0;
+
                     link.a_to_b
-                        .schedule_packet_clearing(height, force_packet_clearing)
+                        .schedule_packet_clearing(Some(height), force_packet_clearing)
                 }
 
+                WorkerCmd::ClearPendingPackets => link.a_to_b.schedule_packet_clearing(None, true),
```
Comment (romac) [inline relayer/src/worker/packet.rs:118]: Is okay to clear pending packets up to the latest height here? ie. whenever we detect that the WS subscription has been closed and that we may have missed events.

File: relayer/src/event/monitor.rs [hunk 2]
```diff
@@ -336,32 +354,55 @@ impl EventMonitor {
...
                     // thus potentially blow up the stack after many restarts.
-                    break;
+                    return Next::Continue;
                 }
             }
         }
+    }
 
-        Next::Continue
+    /// Propagate error to subscribers
```
Comment (adizere) [inline relayer/src/event/monitor.rs:399]: Could we expand on this comment to give some insight as to _who_ are the subscribers?

This code can be difficult to follow because of multiple layers of channels `monitor` <> `runtime` <> `supervisor` <> `workers`. (Not sure how accurate that is, btw.) It would improve understandability (and onboarding others) if we are more concrete here and specify the exact location where this error propagates to.

File: relayer/src/event/monitor.rs [hunk 3]
```diff
@@ -336,32 +354,55 @@ impl EventMonitor {
...
 
             match result {
                 Ok(batch) => self.process_batch(batch).unwrap_or_else(|e| {
-                    warn!(chain.id = %self.chain_id, "{}", e);
+                    error!(chain.id = %self.chain_id, "{}", e);
                 }),
+                Err(Error::SubscriptionCancelled(reason)) => {
+                    error!(chain.id = %self.chain_id, "subscription cancelled, reason: {}", reason);
+
+                    self.propagate_error(Error::SubscriptionCancelled(reason))
```
Comment (adizere) [inline relayer/src/event/monitor.rs:369]: As a minor suggestion: 

Would it improve things if we propagate the error (which triggers packet clearing) __after__ a successful `self.reconnect();`? I can imagine that reconnection could be unsuccessful, or may take some time, multiple tries, and then not sure about what happens with packet clearing.

File: scripts/one-chain [hunk 4]
```diff
@@ -130,8 +130,8 @@ else
   sed -i '' 's#"tcp://127.0.0.1:26657"#"tcp://0.0.0.0:'"$RPC_PORT"'"#g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
   sed -i '' 's#"tcp://0.0.0.0:26656"#"tcp://0.0.0.0:'"$P2P_PORT"'"#g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
   sed -i '' 's#"localhost:6060"#"localhost:'"$PROF_PORT"'"#g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
-  sed -i '' 's/timeout_commit = "5s"/timeout_commit = "1s"/g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
-  sed -i '' 's/timeout_propose = "3s"/timeout_propose = "1s"/g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
+  sed -i '' 's/timeout_commit = "5s"/timeout_commit = "10s"/g' $CHAIN_DIR/$CHAIN_ID/config/config.toml
```
Comment (adizere) [inline scripts/one-chain:133]: Just a note to remember to undo these changes.

File: relayer/src/worker/packet.rs (hunk 1 above)
Comment (adizere) [reply relayer/src/worker/packet.rs:118]: I think so!

**Review by adizere — COMMENTED** [review]
Finished reviewing, I'm still testing though.

File: Cargo.toml [hunk 5]
```diff
@@ -12,9 +12,9 @@ exclude = [
     "proto-compiler"
 ]
 
-# [patch.crates-io]
-# tendermint              = { git = "https://github.com/informalsystems/tendermint-rs", branch = "master" }
-# tendermint-rpc          = { git = "https://github.com/informalsystems/tendermint-rs", branch = "master" }
-# tendermint-proto        = { git = "https://github.com/informalsystems/tendermint-rs", branch = "master" }
-# tendermint-light-client = { git = "https://github.com/informalsystems/tendermint-rs", branch = "master" }
-# tendermint-testgen      = { git = "https://github.com/informalsystems/tendermint-rs", branch = "master" }
+[patch.crates-io]
```
[the remark this answers is not in the document]
Comment (romac) [reply Cargo.toml:15]: Yep

File: relayer/src/event/monitor.rs (hunk 3 above)
Comment (romac) [reply relayer/src/event/monitor.rs:369]: Good point, I was actually wondering about that at some point. I eventually decided to emit the error before the reconnection because it seemed to me that clearing the packets did not require the WebSocket connection to be live. Am I mistaken here?

File: relayer/src/event/monitor.rs (hunk 2 above)
Comment (romac) [reply relayer/src/event/monitor.rs]: Is the expanded comment more satisfactory?

File: relayer/src/event/monitor.rs [hunk 6]
```diff
@@ -336,32 +354,61 @@ impl EventMonitor {
...
                 }
             });
 
             match result {
                 Ok(batch) => self.process_batch(batch).unwrap_or_else(|e| {
-                    warn!(chain.id = %self.chain_id, "{}", e);
+                    error!(chain.id = %self.chain_id, "{}", e);
                 }),
+                Err(Error::SubscriptionCancelled(reason)) => {
+                    error!(chain.id = %self.chain_id, "subscription cancelled, reason: {}", reason);
```
Comment (adizere) [inline relayer/src/event/monitor.rs:367]: I see that the error messages is rendered as follows

> Jul 21 15:25:29.063 ERROR subscription cancelled, reason: Server error: subscription was cancelled (reason: client is not pulling messages fast enough) (code: -32000) chain.id=ibc-0


It may improve log readability if we put the chain id as a tag, just after the `ERROR` message, rendering the line similar to other log output, for instance:

> Jul 21 15:25:29.063 ERROR [ibc-0] subscription cancelled, reason: Server error: subscription was cancelled (reason: client is not pulling messages fast enough) (code: -32000)

**Review by adizere — APPROVED** [review]
Ran different tests and works well, this is a great improvement, great work Romain!

I only had one more minor comment regarding the logs. Beside that, please also remember:

 - changelog
 - revert one-chain temporary changes, revert Cargo.toml patch

**Comment by romac** [comment]
> Ran different tests and works well, this is a great improvement, great work Romain!

Thanks :)

> I only had one more minor comment regarding the logs. Beside that, please also remember:

Just pushed a commit to improve the logs

> * changelog

It was already updated :) https://github.com/informalsystems/ibc-rs/pull/1205/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR29

> * revert one-chain temporary changes

Done

> * revert Cargo.toml patch

Will do once tendermint-rs 0.21 is out and after we do the update in a standalone PR.
